pkg/controller/syncsetinstance/syncsetinstance_controller.go

```diff
@@ -0,0 +1,826 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	hiveresource "github.com/openshift/hive/pkg/resource"
+)
+
+const (
+	controllerName              = "syncsetinstance"
+	unknownObjectFoundReason    = "UnknownObjectFound"
+	unknownObjectNotFoundReason = "UnknownObjectNotFound"
+	applySucceededReason        = "ApplySucceeded"
+	applyFailedReason           = "ApplyFailed"
+	deletionFailedReason        = "DeletionFailed"
+	reapplyInterval             = 2 * time.Hour
+)
+
+// Applier knows how to Apply, Patch and return Info for []byte arrays describing objects and patches.
+type Applier interface {
+	Apply(obj []byte) (hiveresource.ApplyResult, error)
+	Info(obj []byte) (*hiveresource.Info, error)
+	Patch(name types.NamespacedName, kind, apiVersion string, patch []byte, patchType string) error
+}
+
+// Add creates a new SyncSet Controller and adds it to the Manager with default RBAC. The Manager will set fields on the
+// Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return AddToManager(mgr, NewReconciler(mgr))
+}
+
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(mgr manager.Manager) reconcile.Reconciler {
+	r := &ReconcileSyncSetInstance{
+		Client:               hivemetrics.NewClientWithMetricsOrDie(mgr, controllerName),
+		scheme:               mgr.GetScheme(),
+		logger:               log.WithField("controller", controllerName),
+		applierBuilder:       applierBuilderFunc,
+		dynamicClientBuilder: controllerutils.BuildDynamicClientFromKubeconfig,
+	}
+	r.hash = r.resourceHash
+	return r
+}
+
+// applierBuilderFunc returns an Applier which implements Info, Apply and Patch
+func applierBuilderFunc(kubeConfig []byte, logger log.FieldLogger) Applier {
+	var helper Applier = hiveresource.NewHelper(kubeConfig, logger)
+	return helper
+}
+
+// AddToManager adds a new Controller to mgr with r as the reconcile.Reconciler
+func AddToManager(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("syncsetinstance-controller", mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: controllerutils.GetConcurrentReconciles()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to SyncSetInstance
+	err = c.Watch(&source.Kind{Type: &hivev1.SyncSetInstance{}}, &handler.EnqueueRequestForObject{})
+
+	// Watch for changes to ClusterDeployments
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.(*ReconcileSyncSetInstance).handleClusterDeployment),
+	})
```

Comment:
Do we actually need to watch cluster deployments in this controller? I would kind of expect the SyncSet controller to do this, and update the SyncSetInstance if necessary.

Reply:
@dgoodwin the reason that I watch cluster deployments is that whether you can apply a syncsetinstance or not depends on the clusterdeployment status. If the installed status is not true, we need to wait for it to be true (and be notified of that change).